### PR TITLE
[SYCL] Fix group local memory sharing issue

### PR DESCRIPTION
**Pull request by mlychkov**
Group local memory functions depend on inlining pass. When all
optimizations are disabled then they are not inlined. If a user adds
several group local memory calls with the same template argument
then only one group local memory function is specialized and then one
memory buffer is allocated. Thus all variables that save a result of
the group local memory call point to the same buffer.
Enable AlwaysInliner pass even when all llvm passes are disabled as
a workaround.
A More accurate solution with call graph analysis will be added later.

E2E test for this fix is added here: https://github.com/intel/llvm-test-suite/pull/217

### CODE REVIEW

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 1]
```diff
@@ -990,8 +990,14 @@ void EmitAssemblyHelper::EmitAssembly(BackendAction Action,
   }
 
   // Allocate static local memory in SYCL kernel scope for each allocation call.
-  if (LangOpts.SYCLIsDevice)
+  if (LangOpts.SYCLIsDevice) {
+    // Group local memory pass depends on inlining. Turn it on even in case if
+    // all llvm passes or SYCL early optimizations are disabled.
+    // TODO: Remove this workaround when dependency on inlining is eliminated.
```
Comment (AaronBallman) [inline clang/lib/CodeGen/BackendUtil.cpp:996]:
```suggestion
    // FIXME: Remove this workaround when dependency on inlining is eliminated.
```
(Typically, TODO suggests it'll be done as part of the PR, FIXME suggests we'll get it fixed up later.)

**Review by AaronBallman — COMMENTED** [review]
Test coverage for the change?

**Comment by mlychkov** [comment]
> Test coverage for the change?

E2E test for this fix is added here: intel/llvm-test-suite#217

**Comment by mohammadfawaz** [comment]
Is this going to negatively affect targets that rely on early optimizations being off?

**Comment by AaronBallman** [comment]
> > Test coverage for the change?
> 
> E2E test for this fix is added here: [intel/llvm-test-suite#217](https://github.com/intel/llvm-test-suite/pull/217)

FE changes should be tested with FE tests rather than only with end to end tests, shouldn't they? (FE developers will more quickly notice when they regress behavior because they're going to run the FE tests when making changes but aren't likely to run the E2E tests as often.)

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 2]
```diff
@@ -990,8 +990,14 @@ void EmitAssemblyHelper::EmitAssembly(BackendAction Action,
   }
 
   // Allocate static local memory in SYCL kernel scope for each allocation call.
-  if (LangOpts.SYCLIsDevice)
+  if (LangOpts.SYCLIsDevice) {
+    // Group local memory pass depends on inlining. Turn it on even in case if
```
Comment (MrSidims) [inline clang/lib/CodeGen/BackendUtil.cpp:994]: If I understand correctly, always inliner already works when optimizations are disabled, see in this file:
 ```
665   if (CodeGenOpts.OptimizationLevel <= 1) {
 666     bool InsertLifetimeIntrinsics = ((CodeGenOpts.OptimizationLevel != 0 &&
 667                                       !CodeGenOpts.DisableLifetimeMarkers) ||
 668                                      LangOpts.Coroutines);
 669     PMBuilder.Inliner = createAlwaysInlinerLegacyPass(InsertLifetimeIntrinsics);
 670   } else {

```

**Review by MrSidims — COMMENTED** [review]
> Is this going to negatively affect targets that rely on early optimizations being off?

From my understanding, _**always**_ inliner pass already works, when optimizations are disabled.

File: clang/lib/CodeGen/BackendUtil.cpp (hunk 2 above)
Comment (AlexeySachkov) [reply clang/lib/CodeGen/BackendUtil.cpp:994]: @mlychkov, I suggest you to take a look at [clang/test/CodeGenSYCL/kernel-early-optimization-pipeline.cpp](https://github.com/intel/llvm/blob/sycl/clang/test/CodeGenSYCL/kernel-early-optimization-pipeline.cpp) LIT test - this is the way you should be able to write a unit-test for the change, plus you can double-check the information from Dmitry at the same time

Comment (mlychkov) [reply clang/lib/CodeGen/BackendUtil.cpp:994]: > If I understand correctly, always inliner already works when optimizations are disabled, see in this file:
> 
> ```
> 665   if (CodeGenOpts.OptimizationLevel <= 1) {
> 666     bool InsertLifetimeIntrinsics = ((CodeGenOpts.OptimizationLevel != 0 &&
> 667                                       !CodeGenOpts.DisableLifetimeMarkers) ||
> 668                                      LangOpts.Coroutines);
> 669     PMBuilder.Inliner = createAlwaysInlinerLegacyPass(InsertLifetimeIntrinsics);
> 670   } else {
> ```

Please, notice check at the beginning of that function:
```
636 void EmitAssemblyHelper::CreatePasses(legacy::PassManager &MPM,
637                                       legacy::FunctionPassManager &FPM) {
638   // Handle disabling of all LLVM passes, where we want to preserve the
639   // internal module before any optimization.
640   if (CodeGenOpts.DisableLLVMPasses)
641     return;
```

Comment (mlychkov) [reply clang/lib/CodeGen/BackendUtil.cpp:994]: @AlexeySachkov Thank you for the tip.
Added FE test for the fix.

**Comment by mlychkov** [comment]
> > Is this going to negatively affect targets that rely on early optimizations being off?
> 
> From my understanding, _**always**_ inliner pass already works, when optimizations are disabled.

It isn't actually. And this is the cause of the issue with one instance of global variable for two calls of group_local_memory<> with the same template parameter type.

I'm not 100% sure that it won't affect anything, but effect will be only in inlining device functions that are marked with always_inline attribute. If they are marked with this attribute then this is an expected result of a compilation and it shouldn't break correctness of a program.

**Comment by mlychkov** [comment]
It was found out that enabling of AlwaysInliner pass may affect some targets. I set this PR as draft temporary until we find more robust solution.

**Review by mohammadfawaz — APPROVED** [review]

**Comment by mlychkov** [comment]
@elizabethandrews @premanandrao Could you please review this fix?

**Comment by AaronBallman** [comment]
@mlychkov -- I'm a bit confused. https://github.com/intel/llvm/pull/3489#issuecomment-814819946 said you were looking for a more robust solution, a bunch of changes were made, and then reverted back to the original state in https://github.com/intel/llvm/pull/3489/commits/5e3e47ff71ed32566ab7a37b7910b3a3bc59cbc2. Was the more robust solution not needed?

**Comment by mlychkov** [comment]
> @mlychkov -- I'm a bit confused. [#3489 (comment)](https://github.com/intel/llvm/pull/3489#issuecomment-814819946) said you were looking for a more robust solution, a bunch of changes were made, and then reverted back to the original state in [5e3e47f](https://github.com/intel/llvm/commit/5e3e47ff71ed32566ab7a37b7910b3a3bc59cbc2). Was the more robust solution not needed?

Sorry for a confusion. We discussed possible alternative solutions and agreed that the first fix is the most appropriate for now because it doesn't require additional changes on backends side.

**Review by AaronBallman — APPROVED** [review]
Thank you for the explanation, then the changes LGTM.

**Pull request merged**